File: stackops.py
# -*- coding: utf-8 -*-
"""
Created on Thu Jun 30 13:57:12 2022

@author: fc53519
"""
import copy
from ops import *


def is_ascendingP(stack):
    return sorted(stack) == stack

def is_descendingP(stack):
    return sorted(stack, reverse=True) == stack

def set_ascending(stack, verbChar=""):
    """[1, 2, 3]"""
    moves = 0
    #TODO: use rr if more efficient
    if stack.index(max(stack)) > len(stack) // 2:
        while stack[0] != min(stack):
            rr(stack, verbChar)
            moves += 1

    else:
        while stack[0] != min(stack):
            r(stack, verbChar)
            moves += 1
    return moves


def set_descending(stack, verbChar=""):
    """[3, 2, 1]"""
    moves = 0
    #TODO: use rr if more efficient
    if stack.index(max(stack)) > len(stack) // 2:
        while stack[0] != max(stack):
            rr(stack, verbChar)
            moves += 1
    else:
        while stack[0] != max(stack):
            r(stack, verbChar)
            moves += 1
    return moves


def drain(giver, receiver, verbChar=""):
    """Drains all nums in giver to receiver with p"""
    moves = 0
    moves += set_descending(giver)
    while giver:
        p(receiver, giver, verbChar)
        moves += 1
    print(receiver)
    print(giver)
    return moves

def turnedP(stack):
    """Returns whether the stack is Turned or not"""
    if len(stack) < 2:
        return False
    exceptions = 0;
    if stack[0] < stack[-1]:
        return False
    i = 0;
    while ( i < len(stack) - 1):
        if stack[i] > stack[i + 1]:
            exceptions += 1;
            if exceptions > 1:
                return False
        i += 1;
    return True

def miniswitchedP(stack, verifierF):
    """Checks whether a single swap
        orders the stack in the order
        specified by verifierF
        eg: verifierF can be is_ascendingP"""
    s(stack)
    ordered = verifierF(stack)
    s(stack) # undoes the swap
    return ordered



def switchedP(stack, verifierF):
    """A switched stack is in the order specified by
        verifierF with a single swap"""
    if len(stack) < 2:
        return (-1)
    replica = copy.deepcopy(stack)
    index = 0
    rotates = 0
    while index < len(replica):
        rotates = index
        while rotates:
            r(replica)
            rotates -= 1

        s(replica)

        while rotates < index:
            rr(replica)
            rotates += 1

        if verifierF(replica):
            return index
        
        # Goes back to undo the swap
        while rotates:
            r(replica)
            rotates -= 1

        s(replica)

        while rotates < index:
            rr(replica)
            rotates += 1
        index += 1
    return (-1)
    
def swap_at(stack, index, verbChar=""):
    #TODO: Can be more optimized!!! rr might be better
    # than r and vice versa
    if not stack and not index:
        raise Exception()
    moves = 0
    rotates = index
    while rotates:
        moves += r(stack, verbChar)
        rotates -= 1
    moves += s(stack, verbChar)
    while rotates < index:
        moves += rr(stack, verbChar)
        rotates += 1
    # s(stack[index:]) in C
    return moves

# def set_orderb_for(newnum):
#     # Can be a LOT more optimized
#     set_ascending(stackb)
#     while newnum > stackb[0]:
#         rb()
#     pb()

def prepTop(stack, chunks=5):
    """Rotates the given stack to get the more efficient
        lower number to the top"""
    # No negative numbers!
    smallest = min(stack)
    biggest = max(stack)
    chunkSize = (biggest - smallest + chunks) // chunks

    lookChunk = 0
    while lookChunk < chunks:
        found = False
        topi = 0
        while topi < len(stack): 
            if (stack[topi] >= smallest + chunkSize * lookChunk and 
                stack[topi] < smallest + chunkSize * (lookChunk + 1)):
                found = True
                break
            topi += 1

        boti = len(stack) - 1
        while boti >= 0:
            if (stack[boti] >= smallest + chunkSize * lookChunk and 
                stack[boti] < smallest + chunkSize * (lookChunk + 1)):
                found = True
                break
            boti -= 1

        if found:
            boti = len(stack) - boti
            if topi <= boti:
                rotate(stack, topi, "?")
            else:
                revRotate(stack, boti, "?")
            return True

        lookChunk += 1
    # Shouldn't happen
    return False

if __name__ == "__main__":
    a = [75, 68, 100, 43, 23, 2, 98, 77]
    print(prepTop(a))
    print(a)

File: src/stackops.c
#include "../header/push_swap.h"

int	is_ascendingP(t_list *stack)
{
	int		before;

	if (!stack)
		return (-1);

	before = *(int *)(stack->content);
	while (stack->next != NULL)
	{
		stack = stack->next;
		if (before > *(int *)(stack->content))
			return (0);
		before = *(int *)(stack->content);
	}
	return (1);
}


int	is_descendingP(t_list *stack)
{
	int		before;

	if (!stack)
		return (-1);

	before = *(int *)(stack->content);
	while (stack->next != NULL)
	{
		stack = stack->next;
		if (before < *(int *)(stack->content))
			return (0);
		before = *(int *)(stack->content);
	}
	return (1);
}


/*
 * [1, 2, 3]
 */
int	set_ascending(t_list **stack)
{
	int	moves;
	
	moves = 0;
	if (ft_lstindex(*stack, ft_lstmax(*stack)) > (ft_lstlen(*stack) / 2))
	{
		while (**(int **)((*stack)->content) != ft_lstmin(*stack))
		{
			rr(stack);
			moves++;
		}
	}
	else
	{
		while (**(int **)((*stack)->content) != ft_lstmin(*stack))
		{
			r(stack);
			moves++;
		}
	}
	return moves;
}

/*
 * [3, 2, 1]
 */
int	set_descending(t_list **stack)
{
	int	moves;
	
	moves = 0;
	if (ft_lstindex(*stack, ft_lstmax(*stack)) > (ft_lstlen(*stack) / 2))
	{
		while (**(int **)((*stack)->content) != ft_lstmax(*stack))
		{
			rr(stack);
			moves++;
		}
	}
	else
	{
		while (**(int **)((*stack)->content) != ft_lstmax(*stack))
		{
			r(stack);
			moves++;
		}
	}
	return moves;
}


/*
 * Drains all nums in giver to receiver with p
 */
int	drain(t_list **giver, t_list **receiver)
{
	size_t	moves;
	
	moves = set_descending(giver);
	while (*giver)
	{
		p(receiver, giver);
		moves++;
	}
	return moves;
}


/*
 * Returns whether the stack is Turned or not
 */
int	turnedP(t_list *stack)
{
	int	exceptions;
	size_t	i;

	if (ft_lstlen(stack) < 2)
		return (0);
	if (ft_lstget_item(stack, 0) < ft_lstget_item(stack, -1))
		return (0);
	exceptions = 0;
	i = 0;
	while (i < ft_lstlen(stack) - 1)
	{
		if (ft_lstget_item(stack, i) > ft_lstget_item(stack, i + 1))
		{
			exceptions++;
			if (exceptions > 1)
				return (0);
		}
		i++;
	}
	return (1);
}


/*
 * A switched stack is in the order specified by
 * verifierF with a single swap. VerifierF is a predicate
 */
int	switchedP(t_list *stack, int (*f)(t_list *))
{
	t_list	*replica;
	size_t		index, rotates;

	if (ft_lstlen(stack) < 2)
		return (-1);
	replica = ft_lstcpy(stack);
	index = 0;
	rotates = 0;
	while (index < ft_lstlen(replica))
	{
		rotates = index;
		while (rotates)
		{
			r(&replica);
			rotates--;
		}
		s(&replica);
		while (rotates < index)
		{
			rr(&replica);
			rotates++;
		}
		if (f(replica))
			return (index);
		
		// Goes back to undo the swap
		while (rotates)
		{
			r(&replica);
			rotates--;
		}

		s(&replica);

		while (rotates < index)
		{
			rr(&replica);
			rotates++;
		}

		index++;
	}
	return (-1);
}

/*
 * Swaps at given location
 */
int	swap_at(t_list *stack, int index)
{
	int	moves, rotates;

	if (!stack && !index)
		return (-1);
	moves = 0;
	rotates = index;
	while (rotates)
	{
		moves = moves + r(&stack);
		rotates--;
	}
	moves = moves + s(&stack);
	while (rotates < index)
	{
		moves = moves + rr(&stack);
		rotates++;
	}
	return (moves);
}


/*
 * Rotates the given stack to get the more efficient
 * lower number to the top
 */
size_t	prepTop(t_list *stack)
{
	size_t		smallest, biggest, chunkSize, chunks, topi;
	ssize_t		found, boti;
	size_t	moves, lookChunk;

	chunks = 5;
	smallest = ft_lstmin(stack);
	biggest = ft_lstmax(stack);
	chunkSize = (biggest - smallest + chunks) / chunks;

	moves = 0;

	lookChunk = 0;
	while (lookChunk < chunks)
	{
		found = 0;
		topi = 0;
		while (topi < ft_lstlen(stack))
		{
			if ((size_t)ft_pslstget_it(stack, topi) >= smallest + chunkSize * lookChunk &&
					(size_t)ft_pslstget_it(stack, topi) < smallest + chunkSize * (lookChunk + 1))
			{
				found = 1;
				break;
			}
			topi++;
		}
		boti = ft_lstlen(stack) - 1;
		while (boti >= 0)
		{
			if ((size_t)ft_pslstget_it(stack, boti) >= smallest + chunkSize * lookChunk &&
					(size_t)ft_pslstget_it(stack, boti) < smallest + chunkSize * (lookChunk + 1))
			{
				found = 1;
				break;
			}
			boti--;
		}

		if (found)
		{
			boti = ft_lstlen(stack) - boti;
			if ((long long int)topi <= (long long int)boti)
				moves = moves + rotate(&stack, topi);
			else
				moves = moves + revRotate(&stack, boti);
			return (moves);
		}
		lookChunk++;
	}
	// Shouldnt happen
	ft_printf("Shouldnt Be Appearing");
	return (moves);
	
}

/*
int	main(void)
{
	t_list	*stack1;
	// t_list	*stack2;
	t_list	el;
	int		counter;
	int		n[4] = {1,2,3,4};

	counter = 0;
	stack1 = NULL;

	el.content = (void *)(&n);
	el.next = NULL;

	
	stack1 = &el;

	ft_lstadd_back(&stack1, ft_lstnew((void *)&(n[1])));

	ft_lstadd_back(&stack1, ft_lstnew((void *)&(n[2])));

	ft_lstadd_back(&stack1, ft_lstnew((void *)&(n[3])));

	// ft_lstprint(stack1);
	// ft_printf("%d\n", *(int *)(stack1->next->next->next->content));

	// Check asc/desc functions
	ft_printf("%d\n", is_ascendingP(stack1));
	ft_printf("%d\n", is_descendingP(stack1));
}
*/
